### ISSUE TITLE
WPCS flagging an echoed $_GET variable as not-sanitised, while it shouldn't be sanitised, rather escaped

### ISSUE THREAD

**Issue by TukuToi** [WordPress/WordPress-Coding-Standards#2057]
## Bug Description
I have this code:
`<input type="text" placeholder="Hit Enter To Search..." name="s" value="<?php echo isset( $_GET['s'] ) ? esc_html( $_GET['s'] ) : ''; ?>">`

Running WPCS complains that I have an unsanitized $_GET variable.
I do not understand why this would be flagged. It is _echoed_, not stored, so we shouldn't sanitize, we should escape.

Either I have to take a break, or something is wrong here.

I am sorry if I don't see the tree in the forest...

**Comment by dingo-d** [comment]
Did you check the [wiki ](https://github.com/WordPress/WordPress-Coding-Standards/wiki/Fixing-errors-for-input-data) that explains the handling of the input data?

You could see the solution there 😉

**Comment by TukuToi** [comment]
That wiki shows how to _sanitize_
As far I know when we _echo_ we do not sanitize, we escape.
Yet.... that wiki does not touch this aspect. Sanitising is the wrong thing to echo something.

In fact, if I split the $_GET into key/value and then echo the value... it asks me to escape it, not sanitize it.

I am lost 🤣

**Comment by dingo-d** [comment]
Running your code against the `WordPress` codebase (which you shouldn't do really in your project) gives me the following security warnings and errors:

```bash
 1 | WARNING | Processing form data without nonce verification. (WordPress.Security.NonceVerification.Recommended)
 1 | WARNING | Processing form data without nonce verification. (WordPress.Security.NonceVerification.Recommended)
 1 | ERROR   | $_GET data not unslashed before sanitization. Use wp_unslash() or similar
   |         | (WordPress.Security.ValidatedSanitizedInput.MissingUnslash)
 1 | ERROR   | Detected usage of a non-sanitized input variable: $_GET['s'] (WordPress.Security.ValidatedSanitizedInput.InputNotSanitized)
```

Nonce verification is necessary to guard [against CSRF](https://developer.wordpress.org/plugins/security/nonces/) attacks. You need to sanitize and unslash your `$_GET` variable because you are using it in the `input` element. Which is a search field, and that means you are executing a DB query.

So the _correct_ way to write the above code would be:

```php
<?php

if (
	isset( $_GET['s'] )
	&& wp_verify_nonce( sanitize_text_field( wp_unslash( $_GET['s'] ) ), 'search_action' ) ) {
		$search_query = sanitize_text_field( wp_unslash( $_GET['s'] ) );
	?>
		<input type="text" placeholder="Hit Enter To Search..." name="s" value="<?php echo esc_html( $search_query ); ?>">

	<?php
}
```

You might need to check on what nonce you'll use, you can generate one inside your search form using [`wp_create_nonce()`](https://developer.wordpress.org/reference/functions/wp_create_nonce/) function.

**Comment by TukuToi** [comment]
It is just a snip from larger code (which doesn’t query the database) but I get it now;
It flash it because it’s added to an `Input` which might get resubmitted and then becomes … an input not an output.

got it.

thanks!
